### PR TITLE
Remove the spack.architecture module

### PR DESCRIPTION
**Pull request by alalazo**
The `spack.architecture` module contains an `Arch` class that is very similar to `spack.spec.ArchSpec` but points to platform, operating system and target objects rather than "names". There's a TODO in the class since 2016:

https://github.com/spack/spack/blob/abb0f6e27c45758c37fd45d663214b86413fb4f6/lib/spack/spack/architecture.py#L70-L75

and this PR basically addresses that. Since there are just a few places where the `Arch` class was used, here we query the relevant platform objects where they are needed directly from `spack.platforms`. This permits to clean the code from vestigial logic.

Modifications:
- [x] Remove the `spack.architecture` module and replace its use by `spack.platforms`
- [x] Remove unneeded tests

### CODE REVIEW

**Comment by alalazo** [comment]
@spackbot run pipeline

File: lib/spack/spack/build_environment.py [hunk 1]
```diff
@@ -817,8 +820,8 @@ def setup_package(pkg, dirty, context='build'):
         # unnecessary
         module('unload', 'cray-libsci')
```
Comment (haampie) [inline lib/spack/spack/build_environment.py:817]: This is such a weird line...

Comment (haampie) [reply lib/spack/spack/build_environment.py:817]: Way too specific :p

Comment (haampie) [reply lib/spack/spack/build_environment.py:817]: Basically all these lines above run on macos / non-cray-linux, where they make no sense whatsoever

**Review by haampie — DISMISSED** [review]
I don't see any issues going through this

**Comment by alalazo** [comment]
@spackbot run pipeline

**Comment by alalazo** [comment]
@spackbot run pipeline

**Comment by haampie** [comment]
There are no other rebase changes apart from `_on_cray` since last time, right?

**Review by haampie — APPROVED** [review]

**Comment by alalazo** [comment]
No, all the rebase I did were due to file conflicts because other PRs wrote to the same lines - so no new change in behavior.

File: lib/spack/spack/build_environment.py [hunk 2]
```diff
@@ -148,9 +148,10 @@ def __call__(self, *args, **kwargs):
 
 
 def _on_cray():
-    hostarch = arch.Arch(arch.platform(), 'default_os', 'default_target')
-    on_cray = str(hostarch.platform) == 'cray'
-    using_cnl = re.match(r'cnl\d+', str(hostarch.os))
+    host_platform = spack.platforms.host()
```
Comment (tgamblin) [inline lib/spack/spack/build_environment.py:151]: This change makes me happy. `arch.Arch('blah', 'blah', 'blah')` was awful

**Review by tgamblin — APPROVED** [review]

**Comment by alalazo** [comment]
Thanks!

**Comment by haampie** [comment]
@alalazo / @tgamblin this caused a severe performance / startup regression for me with spack shell support. It takes ~ 5 seconds now. Bisected to here.

```
$ git checkout refs/bisect/bad
$ time ./setup-env.sh 
real	0m5.339s
$ git checkout HEAD~1
$ time ./setup-env.sh 
real	0m1.595s
```

```
$ time spack --print-shell-vars sh,modules
_sp_sys_type='linux-ubuntu20.04-zen2'
_sp_compatible_sys_types='linux-ubuntu20.04-zen2:linux-ubuntu20.04-zen:linux-ubuntu20.04-x86_64_v3:linux-ubuntu20.04-x86_64_v2:linux-ubuntu20.04-x86_64'
_sp_tcl_roots='/home/harmen/spack/share/spack/modules'
_sp_lmod_roots='/home/harmen/spack/share/spack/lmod'
_sp_module_prefix='not_installed'

real	0m5.410s
user	0m5.215s
sys	0m0.178s
```

it chokes on `_sp_module_prefix` for a couple seconds, only to return 'not_installed'

```
$ git checkout refs/bisect/bad
$ time spack python -c 'spack.store.db.query("environment-modules target=x86_64")'
[]
real	0m5.450s

$ git checkout refs/bisect/bad^1
$ time spack python -c 'spack.store.db.query("environment-modules target=x86_64")'
[]
real	0m1.759s
```

profiler:
```
5035114/5002194    1.591    0.000    3.631    0.000 /home/harmen/spack/lib/spack/external/archspec/cpu/microarchitecture.py:28(_impl)
  5026884    1.109    0.000    1.399    0.000 /home/harmen/spack/lib/spack/external/archspec/cpu/microarchitecture.py:100(__eq__)
  1455064    0.903    0.000    4.151    0.000 /home/harmen/spack/lib/spack/external/archspec/cpu/microarchitecture.py:91(<genexpr>)
 10371078    0.861    0.000    0.865    0.000 {built-in method builtins.isinstance}
    85592    0.634    0.000    6.055    0.000 /home/harmen/spack/lib/spack/external/archspec/cpu/detect.py:279(compatibility_check_for_x86_64)
```

**Comment by alalazo** [comment]
Just read this. I'll have a look at that this afternoon.

**Comment by haampie** [comment]
fwiw compatibility_check_for_x86_64 went up from 50 calls previously to 85592 after this (spack.store.db.query("environment-modules target=x86_64"))

**Comment by alalazo** [comment]
Maybe I didn't "memoize" a function while moving it. Where is this reproducible? This is what I get on local SSD disk:
```
$ git checkout develop
$ time . share/spack/setup-env.sh 

real	0m0,472s
user	0m0,347s
sys	0m0,116s

$ git checkout 232086c6af2301bcd329b8b857b67284cfb4f2b2
Note: switching to '232086c6af2301bcd329b8b857b67284cfb4f2b2'.
[ ... ]

$ time . share/spack/setup-env.sh 

real	0m0,470s
user	0m0,400s
sys	0m0,062s
```
I'll try to check the number of calls as a metric.

**Comment by tgamblin** [comment]
On my Mac I see it if I just pull develop and launch a shell. Did you try it in a new shell without already having sourced the setup script?

**Comment by alalazo** [comment]
I don't see such high changes in walltime on Daint's /home directory. This is what I am doing:
```console
~> cd github/spack/
~> git checkout develop
~> time . share/spack/setup-env.sh 

real	0m1,853s
user	0m0,652s
sys	0m0,227s
```
If I logout and then:
```console
~> cd github/spack/

~> git checkout 232086c6af2301bcd329b8b857b67284cfb4f2b2
Note: switching to '232086c6af2301bcd329b8b857b67284cfb4f2b2'.
[ ... ]

~> time . share/spack/setup-env.sh 

real	0m1,124s
user	0m0,466s
sys	0m0,196s
```
I have landed on different login nodes though (`daint103` for the first, `daint106` for the second)

**Comment by alalazo** [comment]
@haampie points out in DM that it seems to be coming from `_add_default_platform` when reading specs. If so it all makes sense, since I am trying on a fresh clone with no specs in it.

**Comment by haampie** [comment]
It seems that _host was not memoized, see #26573

**Comment by alalazo** [comment]
Yeah, that's the one line missing from before

**Pull request merged**